packages/syft/src/syft/core/tensor/smpc/share_tensor.py

```diff
@@ -0,0 +1,156 @@
+        return ShareTensor(rank=self.rank, ring_size=self.ring_size)
+
+    @staticmethod
+    @lru_cache(32)
+    def compute_min_max_from_ring(ring_size: int = 2 ** 64) -> Tuple[int, int]:
+        min_value = (-ring_size) // 2
+        max_value = (ring_size - 1) // 2
+        return min_value, max_value
+
+    """ TODO: Remove this -- we would use generate_przs since the scenario we are testing is that
+    the secret is remotly
+    @staticmethod
+    def generate_shares(secret, nr_shares, ring_size=2 ** 64):
+        from .fixed_precision_tensor import FixedPrecisionTensor
+
+        if not isinstance(secret, (int, FixedPrecisionTensor)):
+            secret = FixedPrecisionTensor(value=secret)
+
+        shape = secret.shape
+        min_value, max_value = ShareTensor.compute_min_max_from_ring(ring_size)
+
+        generator_shares = np.random.default_rng()
+
+        random_shares = []
+        for i in range(nr_shares):
+            random_value = generator_shares.integers(
+                low=min_value, high=max_value, size=shape
+            )
+            fpt_value = FixedPrecisionTensor(value=random_value)
+            random_shares.append(fpt_value)
+
+        shares_fpt = []
+        for i in range(nr_shares):
+            if i == 0:
+                share = value = random_shares[i]
+            elif i < nr_shares - 1:
+                share = random_shares[i] - random_shares[i - 1]
+            else:
+                share = secret - random_shares[i - 1]
+
+            shares_fpt.append(share)
+
+        # Add the ShareTensor class between them
+        shares = []
+        for rank, share_fpt in enumerate(shares_fpt):
+            share_fpt.child = ShareTensor(rank=rank, value=share_fpt.child)
+            shares.append(share_fpt)
+
+        return shares
+    """
+
+    @staticmethod
+    def generate_przs(
+        value: Optional[Any],
+        shape: Tuple[int],
+        rank: int,
+        nr_parties: int,
+        seed_shares: int,
+    ) -> "ShareTensor":
+        # syft absolute
+        from syft.core.tensor.tensor import Tensor
+
+        if value is None:
+            value = Tensor(np.zeros(shape, dtype=np.int64))
+
+        # TODO: Sending the seed and having each party generate the shares is not safe
+        # Since the parties would know some of the other parties shares (this might not impose a risk
+        # when shares are not sent between parties -- like private addition/subtraction, but it might
+        # impose for multiplication
+        # The secret holder should generate the shares and send them to the other parties
+        generator_shares = np.random.default_rng(seed_shares)
+
+        share = value.child
+        if not isinstance(share, ShareTensor):
+            share = ShareTensor(value=share, rank=rank)
+
+        shares = [
+            generator_shares.integers(
+                low=share.min_value, high=share.max_value, size=shape
+            )
+            for _ in range(nr_parties)
```

Comment:
We generate for all parties and take only specific rank, would all the parties have the same seed.

Reply:
Yep - all parties have the same seed and generate the same "shares", but they keep only specific shares depending on their rank (if you look at line 118)